### ISSUE TITLE
Wrong SQL generated when using implicit joins in union subquery

### ISSUE THREAD

**Issue by sLite** [jOOQ/jOOQ#9961]
We have found the following regression on query generation for postgres while migrating from JOOQ 3.11 to 3.13.

```
@Test
public void breaks() {
    final var lookup = create
        .select(a.ALARM_ID).from(a)
        .unionAll(
            create
                .select(a.ALARM_ID).from(a)
                .where(a.alarm().EXTERNAL_ALARM_ID.eq("ABC"))
                .and(a.alarm().EXTERNAL_MONITOR_ID.eq("CDE"))
        )
        ;
    assertThat(lookup.fetch()).isNotNull();
}
```

a.alarm() being a nullable self join to a parent record.

### old working query (3.11.11):

```
(
  select "a"."alarm_id"
  from "alarm" as "a"
)
union all (
  select "a"."alarm_id"
  from (
    "alarm" as "a"
      left outer join "alarm" as "alias_96373448"
      on "a"."root_alarm_id" = "alias_96373448"."alarm_id"
    )
  where (
    "alias_96373448"."external_alarm_id" = 'ABC'::citext
    and "alias_96373448"."external_monitor_id" = 'CDE'::citext
  )
)
```

### new invalid SQL (3.13.1):

```
select "a"."alarm_id"
from "alarm" as "a"
union all
select "a"."alarm_id"
from (
  "alarm" as "a"
    left outer join "alarm" as "alias_109520523"
      on "a"."root_alarm_id" = "alias_109520523"."alarm_id"
  ) "a"
    left outer join "alarm" as "alias_109520523"
      on "a"."root_alarm_id" = "alias_109520523"."alarm_id"
  )
where (
  "alias_109520523"."external_alarm_id" = 'ABC'::citext
  and "alias_109520523"."external_monitor_id" = 'CDE'::citext
)
```

We also tried with 3.12.4, this looks different but also generates bad grammar

### new invalid SQL (3.12.4)
```
(
  select "a"."alarm_id"
  from "alarm" as "a"
)
union all (
  select "a"."alarm_id"
  from (
    "alarm" as "a"
      left outer join "alarm" as "alias_96373448"
        on "a"."root_alarm_id" = "alias_96373448"."alarm_id"
    )as "a"
      left outer join "alarm" as "alias_96373448"
        on "a"."root_alarm_id" = "alias_96373448"."alarm_id"
    )
  where (
    "alias_96373448"."external_alarm_id" = 'ABC'::citext
    and "alias_96373448"."external_monitor_id" = 'CDE'::citext
  )
)
```

**Comment by lukaseder** [comment]
Thank you very much for your analysis. The 3.12 version looks similar to https://github.com/jOOQ/jOOQ/issues/9674, I'll investigate that separately

**Comment by lukaseder** [comment]
I can reproduce the problem. It doesn't happen when only the second union subquery uses the implicit join syntax. This seems to work as expected:

```java
create
  .select(a.ALARM_ID).from(a)
  .where(a.alarm().EXTERNAL_ALARM_ID.eq("ABC"))
  .and(a.alarm().EXTERNAL_MONITOR_ID.eq("CDE"))
```

A workaround would be not to use implicit join for the time being.

**Comment by sLite** [comment]
Thank you for the quick investigation and workaround proposal!

But i think you posted the wrong snipped, you reposted my example code ;)

**Comment by sLite** [comment]
Refactored to explicit join, works like a charm. Thanks a lot! :)

**Comment by lukaseder** [comment]
> But i think you posted the wrong snipped, you reposted my example code ;)

What I meant by this is to use the subquery without union all and it works

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14.

I won't backport the fix to 3.13 and earlier versions:

- The fix is rather delicate and depends on a variety of improvements on defining how we internally handle subqueries since 3.14.
- We've added numerous additional integration tests in 3.14 only, for implicit joins, including using implicit joins heavily in `jOOQ-meta`. Backporting those tests is not an option. Without the tests, backporting the delicate change is too risky.
